pkg/common/scrubber.go

```diff
@@ -0,0 +1,341 @@
+	}
+
+	// unmarshal secret into map
+	encodedSecretMap := map[string]string{}
+	if err := json.Unmarshal(secretContentStr, &encodedSecretMap); err != nil {
+		return nil, errors.Wrap(err, "Failed to unmarshal function secret")
+	}
+
+	// decode secret keys and values
+	// convert values to byte array for decoding purposes
+	secretMap, err := s.DecodeSecretData(MapStringStringToMapStringBytesArray(encodedSecretMap))
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to decode function secret data")
+	}
+
+	return secretMap, nil
+}
+
+// DecodeSecretData decodes the keys of a secrets map
+func (s *AbstractScrubber) DecodeSecretData(secretData map[string][]byte) (map[string]string, error) {
+	decodedSecretsMap := map[string]string{}
+	for secretKey, secretValue := range secretData {
+		if secretKey == SecretContentKey {
+
+			// when the secret is created, the entire map is encoded into a single string under the "content" key
+			// which we don't care about when decoding
+			continue
+		}
+		decodedSecretKey, err := s.DecodeSecretKey(secretKey)
+		if err != nil {
+			return nil, errors.Wrap(err, "Failed to decode secret key")
+		}
+		decodedSecretsMap[decodedSecretKey] = string(secretValue)
+	}
+	return decodedSecretsMap, nil
+}
+
+// GenerateFunctionSecretName generates a secret name for a function, in the form of:
+// `nuclio-secret-<project-name>-<function-name>-<unique-id>`
+func (s *AbstractScrubber) GenerateFunctionSecretName(functionName string) string {
+	secretName := fmt.Sprintf("%s-%s", "nuclio", functionName)
+	if len(secretName) > KubernetesDomainLevelMaxLength-8 {
+		secretName = secretName[:KubernetesDomainLevelMaxLength-8]
+	}
+
+	// remove trailing non-alphanumeric characters
+	secretName = strings.TrimRight(secretName, "-_")
+
+	// add a unique id to the end of the name
+	secretName = fmt.Sprintf("%s-%s", secretName, GenerateRandomString(8, SmallLettersAndNumbers))
+
+	return secretName
+}
+
+// GenerateFlexVolumeSecretName generates a secret name for a flex volume, in the form of:
+// `nuclio-flex-volume-<volume-name>-<unique-id>`
+func (s *AbstractScrubber) GenerateFlexVolumeSecretName(functionName, volumeName string) string {
+	secretName := fmt.Sprintf("%s-%s-%s", NuclioFlexVolumeSecretNamePrefix, functionName, volumeName)
+
+	// if the secret name is too long, drop the function and project name
+	if len(secretName) > KubernetesDomainLevelMaxLength {
+		secretName = fmt.Sprintf("%s-%s", NuclioFlexVolumeSecretNamePrefix, volumeName)
+
+	}
+
+	// if the secret name is still too long, trim it and keep space for the unique id
+	if len(secretName) > KubernetesDomainLevelMaxLength-8 {
+		secretName = secretName[:KubernetesDomainLevelMaxLength-8]
+	}
+
+	// remove trailing non-alphanumeric characters
+	secretName = strings.TrimRight(secretName, "-_")
+
+	// add a unique id to the end of the name
+	secretName = fmt.Sprintf("%s-%s", secretName, GenerateRandomString(8, SmallLettersAndNumbers))
+
+	return secretName
+}
```

Comment:
These are specific to the function config scrubber, and are not common with the API gateway scrubber, so they shouldn't belong to the Abstract.

Also, the API gateway scrubber should have a similar function `GenerateAPIGatewaySecretName` or similar.

Reply:
@TomerShor I've generalised name `GenerateFunctionSecretName` to `GenerateObjectSecretName`. Also, `GenerateFlexVolumeSecretName` moved to function config scrubber